### PR TITLE
[CI] Add torchtitan tests to PyTorch CI

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #175901

torchtitan currently runs its own CI against PyTorch nightly every 6-12h,
but PyTorch CI has no awareness of torchtitan. A breaking PyTorch commit
is only discovered hours later with no signal back to the committer. This
adds non-blocking torchtitan tests that run on push to main and on-demand
via ciflow labels.

The dispatch chain is: torchtitan.yml → _linux-build.yml → _linux-test.yml
→ test.sh (pattern-match TEST_CONFIG == *torchtitan*) → lumen_cli →
TorchtitanTestRunner → torchtitan's own test runners.

The lumen_cli runner is simpler than VllmTestRunner mainly because
torchtitan is pure Python and doesn't need a separate build step—torch is
pre-installed from _linux-build.yml. torchao is installed from nightly
since torchtitan depends on it at runtime. The test library YAML delegates
to torchtitan's own test runners (pytest for unit tests, run_tests.py for
integration), so torchtitan owns its test definitions and the daily pin
bump picks up changes automatically.

Test plan:
To trigger the workflow on a PR, add the "ciflow/torchtitan" label via the
GitHub UI. The ciflow bot (configured in pytorch-probot.yml) will
automatically push a git tag ciflow/torchtitan/<pr-number>, which triggers
the torchtitan.yml workflow. After merging, the workflow also runs on every
push to main. workflow_dispatch can be used to trigger it manually from the
Actions tab. Add the label to this PR to validate the end-to-end flow.

Follow-ups for subsequent PRs:
- Golden loss regression checks: torchtitan CI compares training loss
  against golden values; this is omitted here (we just check "doesn't
  crash") and should be added once the basic signal is stable
- Experiment suites: compiler toolkit, SimpleFSDP, VLM, etc. are not
  included yet
- ROCm support: torchtitan supports ROCm but we start CUDA-only

Authored with Claude.

### CODE REVIEW

File: .github/workflows/torchtitan.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+jobs:
+  build:
+    name: torchtitan-x-pytorch-build
+    if: github.repository_owner == 'pytorch'
+    uses: ./.github/workflows/_linux-build.yml
+    with:
+      build-environment: linux-jammy-cuda12.8-py3-gcc11
+      docker-image-name: ci-image:pytorch-linux-jammy-cuda12.8-cudnn9-py3-gcc11
+      cuda-arch-list: '8.0 8.9 9.0 10.0'
+      runner: linux.24xlarge
```
Comment (huydhn) [inline .github/workflows/torchtitan.yml]:
```suggestion
      build-external-packages: 'torchtitan'
      runner: linux.24xlarge.memory
```

Comment (wconstab) [reply .github/workflows/torchtitan.yml]: Thanks @huydhn - trying it!

File: .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# torchtitan Test Library Configuration
+# Each test plan maps to torchtitan's own test runners and pytest suites.
+# When tests are added/removed in torchtitan, the daily pin bump picks them up.
+
+torchtitan_unit_tests:
```
Comment (tianyu-l) [inline .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: I would hope these are not hardcoded in pytorch. Changing them would be a pain.

I feel we could make this coupling softer, e.g. by having a file in torchtitan (e.g. named pytorch_ci_tests.txt to give people a sense) which contains the tests to be run by pytorch CI, so that we can change it freely in torchtitan.

Comment (wconstab) [reply .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: One reason I think it should be this way is that if you edit the test in any way you can push your edit as a PR to PyTorch and verify it worked before landing. If it's in titan, you'd have to land it to titan untested, then wait for the PyTorch pin to advance to include it before you get to see if it worked as expected.

The other reason is it's simpler this way and I think idiomatic for the PyTorch ci config to be defined locally.

Is it really too hard to submit a PyTorch PR for controlling the PyTorch CI?

File: .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    # - test_tokenizer.py, test_download_hf_assets.py: require network access to HuggingFace Hub
+    # - test_dataset_flux.py: requires transformers package (not installed)
+    # - test_flux_config_via_cli: transitively imports transformers via flux config registry
+    - >-
+      pytest tests/unit_tests
+      --ignore=tests/unit_tests/test_tokenizer.py
+      --ignore=tests/unit_tests/test_download_hf_assets.py
+      --ignore=tests/unit_tests/test_dataset_flux.py
+      -k "not test_flux_config_via_cli"
+      --cov=. --durations=20 -vv
```
Comment (fegin) [inline .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: I'm not sure how much signal we can get from unit_tests. afaik, most of the unittest either use mock or very simple module to verify the logic. But this is a CPU test, so it may be okay since CPU resource is not expensive.

File: .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  env_vars:
+    NGPU: "8"
+  steps:
+    # TODO: we should debug this, maybe it's a known issue or a test config issue,
+    # then we should also have a cleaner way to disable tests than using sed (--exclude / --skip flag?)
+    # Disable cpu_offload+opt_in_bwd+TP+DP+CP test (times out on CI runners)
+    - >-
+      sed -i 's/"cpu_offload+opt_in_bwd+TP+DP+CP",/"cpu_offload+opt_in_bwd+TP+DP+CP",\n            disabled=True,/'
+      tests/integration_tests/features.py
+    - python -m tests.integration_tests.run_tests --test_suite features artifacts-to-be-uploaded --ngpu 8
```
Comment (fegin) [inline .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: Curious will this be on H100?

File: .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml (hunk 3 above)
Comment (wconstab) [reply .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: i am happy to delete it if its not useful. it seemed like a good thing to try first to get the test system working.

File: .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml (hunk 4 above)
Comment (wconstab) [reply .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: i didn't check :)
i will discuss with the team before i dig into the details too much.
i think the first thing i want to do is land some small set of tests so we can validate the automatic pin advancing is working and then do a follow up PR to fine-tune what tests run /  when

File: .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml (hunk 2 above)
Comment (tianyu-l) [reply .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: > If it's in titan, you'd have to land it to titan untested, then wait for the PyTorch pin to advance to include it before you get to see if it worked as expected.

hmm this part not clear to me:

Let's say you put it in torchtitan. Then you can use titan CI to verify the change with last pytorch.
pytorch is still using last titan, and when it tries to update pin, it may fail because of titan's latest change. Then what happened in your workflow?

Let's say you put in pytorch. Then after titan test structure change, it won't fail with last pytorch.
pytorch is still using last titan, and when it tries to update pin, it MUST fail because the test structure change. Then what happened? If the pin is not moved forward, a PR to pytorch won't help as it is testing against last titan. Would a user have a submit a PR on test structure in this file, as well as to the pin commit?

Comment (wconstab) [reply .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: > Let's say you put it in torchtitan. Then you can use titan CI to verify the change with last pytorch.

well, i'm assuming we are selecting a subset of torchtitan's tests to run in pytorch CI.  I'm also assuming that _all_ the torchtitan tests, except those that are marked 'skip' or 'disable' in the test code, are running/passing in CI against some version of pytorch. e.g., known working.

This list is specifically "Which subset to we run in particular jobs on pytorch/pytorch CI?  i don't think we'd actually have code on the torchtitan CI side that reads this file or executes the subset of tests from it.  (Since the tests are already covered in torchtitan CI directly).

> it may fail because of titan's latest change. Then what happened in your workflow?

if someone adds a new test in torchtitan, then currently, it may get picked up 'automatically' and run as part of a glob of tests when the nightly pytorch job advances the pin and runs CI on latest titan.  (a) if the test fails, then we'll have to manually debug/intervene to get it fixed. (b) if all tests pass, the new test will silently be included in the torchtitan CI on pytorch.  Unless the new test is not part of one of the existing test suites that we run from in pytorch.

> Let's say you put in pytorch. Then after titan test structure change, it won't fail with last pytorch.
pytorch is still using last titan, and when it tries to update pin, it MUST fail because the test structure change. Then what happened? If the pin is not moved forward, a PR to pytorch won't help as it is testing against last titan. Would a user have a submit a PR on test structure in this file, as well as to the pin commit?

not sure i followed this one. (trying to think of what example you had in mind).
What if you split 'features_test' into 2 sub-test suites.  in this case pytorch CI could fail on the nightly pin advancement if it can not find the old 'features_test' file.  PyTorch would require a PR to update the CI config to point to the new files.  It would be necessary to include the pin bump in this PR so that it tests against the version of titan that it works with.  (Similar to how this PR #175901 manually sets the pin to a 'working version' to get a clean signal so it can land

> I feel we could make this coupling softer, e.g. by having a file in torchtitan (e.g. named pytorch_ci_tests.txt to give people a sense

We could do something like this instead. It would have the advantage that you could update the test structure and update pytorch's calling of the tests in one PR.  But it would also be more complicated to build.  We'd probably need to have several test lists, one per 'job' (e.g. feature test, integration test, model test- which run on different runners), and then add logic to the pytorch yml to parse those files and run the corresponding tests.  Or, we could make new executable scripts in torchtitan that we directly run from pytorch, and the test filtering goes in those scripts.

If you prefer this way, then i think it's a good time to stop and discuss which tests we want to run and how to organize them (nightly, per-PR, etc) and what runners they need for them, so we can get the right thing set up on the torchtitan side first and then call into it from this PR.

Comment (tianyu-l) [reply .ci/lumen_cli/cli/lib/core/torchtitan/torchtitan_test_library.yaml]: > PyTorch would require a PR to update the CI config to point to the new files. It would be necessary to include the pin bump in this PR so that it tests against the version of titan that it works with. (Similar to how this PR https://github.com/pytorch/pytorch/pull/175901 manually sets the pin to a 'working version' to get a clean signal so it can land

Yeah I think this explains it. It is this particular step I'm proposing to save for titan devs.

> If you prefer this way, then i think it's a good time to stop and discuss which tests we want to run and how to organize them (nightly, per-PR, etc) and what runners they need for them, so we can get the right thing set up on the torchtitan side first and then call into it from this PR.

That's fair. I think should at least include
- numerics test
- compiler / sac / flex-attn related tests which constantly break torchtitan

I don't mind us starting with your current approach and iterate. I think a minimal test including everything I mentioned above is the feature test.

File: .github/workflows/torchtitan.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      build-environment: linux-jammy-cuda12.8-py3-gcc11
+      docker-image-name: ci-image:pytorch-linux-jammy-cuda12.8-cudnn9-py3-gcc11
+      cuda-arch-list: '8.0 8.9 9.0 10.0'
+      runner: linux.24xlarge.memory
+      test-matrix: |
+        { include: [
+          { config: "torchtitan_unit_tests", shard: 1, num_shards: 1, runner: "linux.24xlarge" },
+          { config: "torchtitan_features_integration", shard: 1, num_shards: 1, runner: "linux.g5.48xlarge.nvidia.gpu" },
+          { config: "torchtitan_models_integration", shard: 1, num_shards: 1, runner: "linux.g5.48xlarge.nvidia.gpu" },
+          { config: "torchtitan_h100_integration", shard: 1, num_shards: 1, runner: "linux.aws.h100.8" }
```
Comment (wconstab) [inline .github/workflows/torchtitan.yml]: @tianyu-l @fegin let's use this thread to discuss which tests to run in which job.  

**Jobs:**
A job is the unit of scheduling in this design, so if we want separate tests to run on PR vs nightly or periodic, they could be separate jobs.  (maybe there is a smart way to have one job that filters tests differently depending on a variable but I didn't try).

**Runners:**
- cpu - linux.24xlarge
- 8x A10G gpu - linux.g5.48xlarge.nvidia.gpu, supports autoscaling, afaiu
- 8x H100 gpu - linux.aws.h100.8  ** this one we only have a couple of bare-metal instance. we need to be careful not to overuse it, or it'll increase queue time, unless we manually ask for more instances (and $$)

**Tests**
- @tianyu-l mentioned adding numerics / compiler / sac / flex-attn related tests (all contained inside 'features_integration' suite as the MVP
- do we care about h100 or models tests?
- @fegin mentioned maybe we should not care about unit tests

**Proposal A** 
- One job that just runs feature tests on A10G gpu.  (minus the cpu offloading test that is hanging for now, until debugging that)
- The job runs nightly for now for pin advancement, but can also be opted-into by applying the ciflow/torchtitan tag 
- we can put the 'filtering' part into some pytorch_ci_runner script in torchtitan

What do folks think of that / propose instead?

Comment (tianyu-l) [reply .github/workflows/torchtitan.yml]: sounds good to me.

> do we care about h100 or models tests?
> @fegin mentioned maybe we should not care about unit tests

traditionally these are not big headache, so I think we can start with feature test. If we have bandwidth, should also include model test because it has MoE architecture.

**Review by tianyu-l — APPROVED** [review]
Thanks for enabling this!

**Review by fegin — APPROVED** [review]
